pkg/logging/utils/daemonset.go

```diff
@@ -0,0 +1,257 @@
+package utils
+
+import (
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/apps/v1beta2"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/rbac/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	loggingconfig "github.com/rancher/rancher/pkg/logging/config"
+)
+
+func CreateFluentd(k8sClient kubernetes.Interface, namespace string) (err error) {
+	serviceAccount := newFluentdServiceAccount(loggingconfig.FluentdName, namespace)
+	serviceAccount, err = k8sClient.CoreV1().ServiceAccounts(namespace).Create(serviceAccount)
+	defer func() {
+		if err != nil {
+			k8sClient.CoreV1().ServiceAccounts(namespace).Delete(serviceAccount.Name, &metav1.DeleteOptions{})
+		}
+	}()
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	projectRoleBind := newFluentdRoleBinding(loggingconfig.FluentdName, namespace)
+	projectRoleBind, err = k8sClient.RbacV1beta1().ClusterRoleBindings().Create(projectRoleBind)
+	defer func() {
+		if err != nil {
+			k8sClient.RbacV1beta1().ClusterRoleBindings().Delete(projectRoleBind.Name, &metav1.DeleteOptions{})
+		}
+	}()
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	daemonset := newFluentdDaemonset(loggingconfig.FluentdName, namespace, loggingconfig.FluentdName)
+	daemonset, err = k8sClient.AppsV1beta2().DaemonSets(namespace).Create(daemonset)
+	defer func() {
+		if err != nil {
+			k8sClient.AppsV1beta2().DaemonSets(namespace).Delete(daemonset.Name, &metav1.DeleteOptions{})
+		}
+	}()
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+func RemoveFluentd(k8sClient kubernetes.Interface, namespace string) {
```

Comment:
This method should return error. If we decide to ignore it, ignore in the caller stack.

Reply:
Also I'm just curious why we ignore the delete errors. It looks like the cleanup should happen properly, and errors should not be ignored, unless is not found.

Reply:
I afraid when remove the first object (for example service) occur error and return, the other parts will not go on, the role, rolebinding, daemonset still here.

Reply:
@aiwantaozi you can track if the error happened, and return it only at the end. This way we can attempt to cleanup everything. You can use errgroup to do a cleanup in parallel, and return the first error: https://godoc.org/golang.org/x/sync/errgroup#Group.Wait; example of use: https://github.com/rancher/rke/blob/caac0a9927467dc7f114495385d845d16c7d7f4f/services/workerplane.go#L27

It's just we have to return an error instead of just logging it in, so the cleanup will be attempted again, if some part of it failed.